File: sdk/python/src/cigar_sdk/idempotency.py
"""Idempotency-key helpers."""

from __future__ import annotations

import re
import uuid

from cigar_sdk.errors import ValidationError

_PREFIX = re.compile(r"^[A-Za-z0-9._~-]{1,32}$")
_KEY = re.compile(r"^[\x21-\x7e]{1,256}$")


def create_idempotency_key(prefix: str = "cigar") -> str:
    if _PREFIX.fullmatch(prefix) is None:
        raise ValidationError("idempotency prefix must be 1..32 unreserved ASCII characters")
    return f"{prefix}-{uuid.uuid4()}"


def validate_idempotency_key(value: str) -> str:
    if _KEY.fullmatch(value) is None:
        raise ValidationError("idempotency key must be 1..256 visible ASCII characters")
    return value

File: sdk/typescript/src/idempotency.ts
import { randomUUID } from "node:crypto";

import { ValidationError } from "./errors.js";

const KEY_PATTERN = /^[\x21-\x7e]{1,256}$/u;

export function createIdempotencyKey(prefix = "cigar"): string {
  if (!/^[A-Za-z0-9._~-]{1,32}$/u.test(prefix)) {
    throw new ValidationError("idempotency prefix must be 1..32 unreserved ASCII characters");
  }
  return `${prefix}-${randomUUID()}`;
}

export function validateIdempotencyKey(value: string): string {
  if (!KEY_PATTERN.test(value)) {
    throw new ValidationError("idempotency key must be 1..256 visible ASCII characters");
  }
  return value;
}

